Generate full functions for lambdify

caley opened the pull request:
#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

Changes lambdify to generate full functions instead of lambdas for #14702.  This is work in progress, comments and suggestions very welcome. 

#### Brief description of what is fixed or changed
Besides #14702, this PR
  * Makes dummify smarter  #12463, e.g.
```
>>> AAA = sympy.symbols('3A')
>>> print(functiondefstr([x, y, AAA], x * y + AAA))
def _lambdifygenerated(x, y, _Dummy_19):
    return (_Dummy_19 + x*y)
```
Code generation is performed by `functiondefstr` instead of `lambdastr`.  `dummify = True` is still used by default in `lambdify`, but that could be changed.

  * Improves speed of argument unpacking #14654 (#14691)
```
>>> print(functiondefstr([(x, y)], x + y))
def _lambdifygenerated(_Dummy_18):
    [x, y] = _Dummy_18
    return (x + y)
```

  * Generate code for the numpy argument wrapping in place of the decorator #14671 (#14683)
```
>>> f = lambdify([x, y], x + y, modules='numpy', dummify=False)
```
Generates code
```
def _lambdifygenerated(x, y):
    builtin_numerics = (int, long, float, complex)
    if isinstance(x, builtin_numerics): x = array(x)
    if isinstance(y, builtin_numerics): y = array(y)
    return (x + y)
```
Note, it still uses the decorator if any of specified arguments are strings to ensure we get the same behaviour.  Otherwise we'd have to parse the strings in functiondefstr (currently they're just copied into the output).

#### Other comments

Currently `functiondefstr` takes extra optional arguments to control the different code generation options.  This is already a bit messy, and will get worse as more options are added.  Does anyone have suggestions on how to refactor it?

Also `functiondefstr` could be renamed to (and replace) `lambdastr`, but we probably don't want to expose all the code generation options.

caley committed:
Generate full functions in lambdify

Initial refactor of the lambdify code to generate full functions
instead of lambdas (see #14702).  The code generation is performed by
functiondefstr() instead of lambdastr() (lambdastr could potentially
be removed).

Some examples:

>>> from sympy.utilities.lambdify import functiondefstr
>>> from sympy.abc import w, x, y, z
>>> print(functiondefstr((x, y), x * y))
def _lambdifygenerated(x, y):
    return (x*y)
>>> print(functiondefstr([[x, y]], x * y))
def _lambdifygenerated(_Dummy_18):
    [x, y] = _Dummy_18
    return (x*y)

Or to work with tensorflow (see #14655)
>>> print(functiondefstr([[x, y]], x * y, unpackByIndex=True))
def _lambdifygenerated(_Dummy_22):
    [x, y] = [_Dummy_22[0],_Dummy_22[1]]
    return (x*y)

The 'dummification' has also been made smarter:
>>> import sympy
>>> AAA = sympy.symbols('3A')
>>> print(functiondefstr([x, y, AAA], x * y + AAA))
def _lambdifygenerated(x, y, _Dummy_19):
    return (_Dummy_19 + x*y)

caley committed:
Replace lambdify numpy decorator by code generation

The argument wrapping performed by the decorator is now performed
by generated code.  This is to improve on the performance regression

Also, the more complicated code generation has been pulled out into
separate functions, to make it clear what each part does.

asmeurer commented:
This should also make numba work with lambdified functions again. Presently it doesn't work because numba doesn't work with decorators (see https://github.com/numba/numba/issues/2328).

asmeurer commented:
I did some basic tests and it looks like numba works here.

asmeurer commented:
> dummify = True is still used by default in lambdify, but that could be changed.

Yes, let's make the default False, which will dummify only when necessary (or we could make it `None` and make False force disable it, but I don't think that would be useful).

asmeurer commented on sympy/utilities/lambdify.py:
```diff
@@ -588,6 +611,177 @@ def flat_indexes(iterable):
+        #XXX: This has to be done here because of circular imports
+        from sympy.printing.lambdarepr import lambdarepr
+
+    identre = re.compile('^[a-zA-Z_][a-zA-Z0-9_]*$')
```
This should use something like this https://stackoverflow.com/a/29586366/161801

caley committed:
Further changes for lambdify to generate full functions

  * Changed default to dummify=False in lambdify
  * lambdastr returns function definitions instead of lambdas.  Added
    option 'name' to specify the generated function name.

caley commented:
I've made dummify=False by default and used `isidentifier` to check for valid identifiers in Python 3.  

I've also changed `lambdastr` to return the new generated functions, instead of lambdas: both `lambdastr` and `lambdify` call `_lambdastrimpl`, which does the code generation.  I've added an optional parameter `name` to `lambdastr` (for the function name), but haven't exposed the other options in `_lambdastrimpl`. 

To me, this seems the simplest sensible way for `lambdastr` to work.  But if it needs to produce the same code as `lambdify`, we will need to add more controls.   Perhaps adding a `modules` argument, as in `lambdify` would be sufficient.

asmeurer commented:
There are some test failures here.

asmeurer commented:
@bjodah would you have a chance to review this?

bjodah commented:
Apart from the test failures I think it's a good idea to have `lambdify` generating functions.

People might rely on `lambdastr` generating actual lambda functions, e.g., for keyword argument defaults:
`'def generated_func(x, f=%s):' % lambastr(...)` 

besides, a function named `lambdastr` which does not return a lambda function is confusing..
(lambdify might be a confusing name too but let's leave that discussion for another time).

So what about leaving `lambdastr` as is, and introduce another function (e.g. "evaluator_func_str"?) 
and make `lambdify` use the new function?


But perhaps we should introduce anther function (or rename `lambdastr`) because

caley commented:
ok, so I'll restore lambdastr to what it was before.  I won't add a public API to give the 'function definition' string, unless there is a particular use for it?  lambdify inserts the source into the function doc string, so people can at least see it that way.

The test failures are because I didn't have numexpr installed, and didn't check the skipped tests.  I have a simple fix if we make `dummify=True` the default again, but I'll try to think of a better fix.

moorepants commented:
Just to throw my two cents in. I think lambdify is a really poor name. Maybe the change in this PR is a good opportunity to make a new function name.

asmeurer commented:
What would you call it? 

Note that lambdify is one of the most important functions in SymPy and there are a lot of resources out there referencing it. So a name change shouldn't be taken lightly.

caley committed:
Define class structure for lambdify code generation

Moved the code generation into class _EvaluatorPrinter, with sub-classes
for customisations for different modules.  Fixed test failures with
NumExpr, and added a further test case for the dummify code.

Also, restored lambdastr to previous code producing lambda function
definitions.

bjodah reviewed (approved):
Looks really good!

caley committed:
Handle Dummy-type arguments passed to lambdify

Input arguments to lambdify of type Dummy can cause name collisions with
arguments of type Symbol.  Force dummify of all arguments in this case.

caley commented:
I've just added a small change to handle a case where `dummify=False` (now the default) was failing.  It's for the test case `test_lambdify_mixed_symbol_dummy_args()` I added in the previous commit.

asmeurer commented on sympy/utilities/lambdify.py:
```diff
@@ -588,6 +616,276 @@ def flat_indexes(iterable):
+    """Generate numpy argument wrapping code for _lambdastrimpl."""
+    integer_names = ', '.join(cls.__name__ for cls in integer_types)
+
+    lines = ['builtin_numerics = ({}, float, complex)'.format(integer_names)]
```
It would be better to just add this to the function namespace. Otherwise it has to be run every time the function is run.

caley replied:
Actually, all this code was meant to be deleted!  It's now inside the _EvaluatorPrinter class and sub-classes.  But yes, this can be done in the _NumpyEvaluatorPrinter class definition, instead of every function call.  Is it best if I make a new pull request to change this and remove the unused code?

asmeurer replied:
This was merged so yes, you need to create a new pull request.

asmeurer commented:
Regarding #14671, this doesn't bring the performance back where it was because we still do the isinstance checks and call `array` on the input. I think the only way we can get performance like we had without resorting to something like `numba` is to try to find some way to remove the need for converting the input.

smichr commented:
needs a test added for this issue
